### PR TITLE
[REF]: Refactor network.add_tonic_bias

### PR DESCRIPTION
**Pull request by kmilo9999**
- Refactor `network.add_tonic_bias` to accept a dictionary of cell types and their amplitudes.
- Changed how drivers create tonic inputs
- Adjusted unit tests to the new `network.add_tonic_bias` arguments

### CODE REVIEW

File: hnn_core/network.py
Comment (jasmainak) [inline hnn_core/network.py:1135]:
```suggestion
        cell_types_amplitudes : dict
```

File: hnn_core/network.py
Comment (gtdang) [reply hnn_core/network.py:1165]: Discussed and we are keeping this as-is.

File: hnn_core/network.py
Comment (jasmainak) [reply hnn_core/network.py:1135]: This is technically a breaking change ... can we just overload the parameter instead

**Comment by jasmainak** [comment]
Don't forget to add tests!

**Comment by kmilo9999** [comment]
> Don't forget to add tests!
@jasmainak 
I modified the current unit test that use this function. Should I add more?

File: hnn_core/network.py
Comment (gtdang) [reply hnn_core/network.py:1130]: Discussed and we will rename this to "amplitude"

**Review by gtdang — CHANGES_REQUESTED** [review]

File: hnn_core/network.py
Comment (gtdang) [reply hnn_core/network.py:1135]: What do you mean by overload? Do we want to keep the existing functionality with the cell_type argument and single amplitude and have an additional argument for a dictionary specification?

Comment (jasmainak) [reply hnn_core/network.py:1135]: Keep the old functionality but throw a deprecation warning if `cell_type` is not `None`. For old functionality `amplitude` stays float. For new functionality `amplitude` is `dict`. That way you don't break old code that users have already written.

Comment (kmilo9999) [reply hnn_core/network.py]: I refactored the code to keep `add_tonic_bias ` old behavior and support the new one

File: hnn_core/network.py
Comment (gtdang) [inline hnn_core/network.py:1168]:
```suggestion
        amplitude: dict | float
            A dictionary of cell type keys (str) to amplitude values (float).
            Valid inputs for cell types are those listed in `net.cell_types`.
            If `cell_types` is not None, `amplitude` should be
            a float indicating the amplitude of the tonic input
            for the specified cell type.
```

File: hnn_core/network.py
Comment (gtdang) [inline hnn_core/network.py:1153]:
```suggestion
        """Attaches parameters of tonic bias input for given cell types.
        
        Multiple tonic biases can be added by providing a dictionary of cell-type 
        to amplitude key-value pairs to the amplitude keyword. Start and stop times
        will be the same across all tonic biases.
        
        A single tonic bias can be added by supplying a cell-type name and 
        amplitude value to the `cell_type` and `amplitude` keywords, respectively.
```

File: hnn_core/network.py
Comment (gtdang) [inline hnn_core/network.py:1172]:
```suggestion
            simulation). This value will be applied to all the tonic biases if
            multiple are specified with the `amplitude` keyword. 
```

File: hnn_core/network.py
Comment (gtdang) [inline hnn_core/network.py:1176]:
```suggestion
            This value will be applied to all the tonic biases if
            multiple are specified with the `amplitude` keyword.
```

**Review by gtdang — CHANGES_REQUESTED** [review]
Looks pretty clean to me. Added suggestions for the docstring.

File: hnn_core/network.py
Comment (kmilo9999) [reply hnn_core/network.py]: Done

File: hnn_core/network.py
Comment (kmilo9999) [reply hnn_core/network.py]: Done

File: hnn_core/network.py
Comment (kmilo9999) [reply hnn_core/network.py]: Done

File: hnn_core/network.py
Comment (kmilo9999) [reply hnn_core/network.py]: Done

File: hnn_core/network.py
Comment (kmilo9999) [reply hnn_core/network.py]: Done

**Review by gtdang — APPROVED** [review]
Looks good to me! Great job!

File: hnn_core/tests/test_network.py
Comment (ntolley) [inline hnn_core/tests/test_network.py:761]:
```suggestion
        'L2_pyramidal': 0.5,
        'name_nonexistent': 1.0
```

File: hnn_core/tests/test_network.py
Comment (ntolley) [inline hnn_core/tests/test_network.py:769]:
```suggestion
        'L2_pyramidal': 1.0,
        'L5_basket': 0.5
```

Comment (ntolley) [reply hnn_core/tests/test_network.py:769]: Better to test that we're able to add multiple tonic biases simultaneously with this new API

File: hnn_core/tests/test_network.py
Comment (ntolley) [inline hnn_core/tests/test_network.py:799]: While this API is going to be deprecated, we should have at least one test to make sure it still functions until we completely deprecate it

File: hnn_core/network.py
Comment (ntolley) [inline hnn_core/network.py:1183]: can you add a test for this error? this would be an appropriate one to trigger it:
```py
net.add_tonic_bias(cell_type='L5_pyramidal', amplitude={'L5_pyramidal': 0.1},)
```

Comment (ntolley) [reply hnn_core/network.py:1183]: Also we have standard functions to do type and option checks like this, see:
https://github.com/jonescompneurolab/hnn-core/blob/5e04b177b0de84046a5c75ca1cd584455c4dc482/hnn_core/network.py#L1234

https://github.com/jonescompneurolab/hnn-core/blob/5e04b177b0de84046a5c75ca1cd584455c4dc482/hnn_core/network.py#L970



Here are the function definitions:
https://github.com/jonescompneurolab/hnn-core/blob/5e04b177b0de84046a5c75ca1cd584455c4dc482/hnn_core/externals/mne.py#L175

https://github.com/jonescompneurolab/hnn-core/blob/5e04b177b0de84046a5c75ca1cd584455c4dc482/hnn_core/externals/mne.py#L211

File: hnn_core/network.py
Comment (ntolley) [inline hnn_core/network.py:1192]: add test for this warning as well

**Review by ntolley — COMMENTED** [review]
@kmilo9999 this is looking great!

See my comments above, there's a few warnings/errors that need to be tested (also I'd recommend using `_validate_type()` to keep the error messages consistent with the rest of the codebase).

Reading this PR I'm realizing that in a followup PR we should probably add a `name` option so that multiple tonic biases can be defined `net.add_tonic_bias(name='tonic1')`. This will make the API function almost identically as `net.add_evoked_drive()` and other external drives, as well as allow more complex tonic bias patterns.

File: hnn_core/network.py
Comment (jasmainak) [inline hnn_core/network.py:1178]:
```suggestion
        if cell_type is not None:
```

File: hnn_core/network.py
Comment (jasmainak) [inline hnn_core/network.py:1182]:
```suggestion
            if not isinstance(amplitude, float):
```

File: hnn_core/network.py
Comment (jasmainak) [inline hnn_core/network.py:1130]: can you make this a separate function rather than have it as a method ... the more methods you add, the more complicated the `Network` class becomes and harder to refactor in the future.

File: hnn_core/tests/test_network.py
Comment (jasmainak) [reply hnn_core/tests/test_network.py:799]: also we should check that the deprecation warning is raised. Leave a clear comment to indicate which part of the test is for deprecated functionality so it can be deleted.

File: doc/whats_new.rst
Comment (jasmainak) [inline doc/whats_new.rst:90]: this is too complicated for a new user to understand ... can you simplify the description? Something like:

"cell-specific tonic bias can now be provided using the argument `amplitude` in network.add_tonic_bias"

File: hnn_core/network.py
Comment (ntolley) [reply hnn_core/network.py:1182]: Actually I think this should use `_validate_type()` to be consistent with the rest of the codebase

File: hnn_core/network.py
Comment (jasmainak) [reply hnn_core/network.py:1192]: it's a bit of a strange warning to raise ... when would `len(amplitude)` have 0 length?

Comment (gtdang) [reply hnn_core/network.py:1192]: Hm. Maybe from a GUI standpoint it could happen if you add a tonic bias "drive" but don't edit any of the amplitude? Though seems like that warning/logic would be better handled in the GUI and not the API.

Comment (jasmainak) [reply hnn_core/network.py:1192]: yeah, sounds like something that should be handled in the GUI. Perhaps this speaks to Nick's point of having names for biases as well ... but that may change existing code. I would vote for allowing this "bug" as long as the user can see the list of all added biases in the GUI

File: hnn_core/tests/test_network.py
Comment (kmilo9999) [reply hnn_core/tests/test_network.py:762]: Done

File: hnn_core/tests/test_network.py
Comment (kmilo9999) [reply hnn_core/tests/test_network.py]: Done

File: hnn_core/tests/test_network.py
Comment (kmilo9999) [reply hnn_core/tests/test_network.py:799]: I added test to check deprecation warnings

File: hnn_core/network.py
Comment (kmilo9999) [reply hnn_core/network.py]: Done. Changed the `isinstance` for the standard verification functions in hnn

File: hnn_core/network.py
Comment (kmilo9999) [reply hnn_core/network.py]: Done

File: hnn_core/network.py
Comment (kmilo9999) [reply hnn_core/network.py]: Done

File: hnn_core/network.py
Comment (kmilo9999) [reply hnn_core/network.py]: I replaced it as indicated in the comment

File: hnn_core/network.py
Comment (kmilo9999) [reply hnn_core/network.py]: I added extra text indicating what users can do

File: doc/whats_new.rst
Comment (kmilo9999) [reply doc/whats_new.rst]: Done

File: hnn_core/network.py
Comment (gtdang) [inline hnn_core/network.py:1164]: What happens if an `int` is passed in? Will it raise error?

Comment (kmilo9999) [reply hnn_core/network.py:1164]: Indeed, it throws an error. Should we expect `int`  as well?

File: hnn_core/network.py
Comment (kmilo9999) [reply hnn_core/network.py]: I moved it this block of code as a separate function

File: hnn_core/network.py
Comment (kmilo9999) [reply hnn_core/network.py:1164]: @ntolley any thoughts?

**Comment by kmilo9999** [comment]
@ntolley @gtdang I applied the last changed. This one is ready to go!

**Review by ntolley — APPROVED** [review]
Looks great to me @kmilo9999!

File: hnn_core/network.py
Comment (jasmainak) [reply hnn_core/network.py:1172]: this wasn't removed? could you address this in the GUI PR?

File: hnn_core/network.py
Comment (jasmainak) [inline hnn_core/network.py:1160]: probably a bit late but I find this a little cryptic for the user. Would have instead said:

'dictionary to provide cell-specific tonic bias'

**Comment by jasmainak** [comment]
made a couple of last-minute comments @kmilo9999 . Sorry I didn't have time to look before but would be great if you could address in next PR. Thanks for your efforts !

**Comment by ntolley** [comment]
Apologies @jasmainak I was hasty with the merge button! We can definitely address this in the follow up for #715

**Comment by jasmainak** [comment]
No, you're fine! Please don't wait for my green light to merge.

One more thing to check: was any example updated? Maybe the gamma example uses tonic bias?

**Pull request merged**
